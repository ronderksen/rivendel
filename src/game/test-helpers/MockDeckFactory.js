function createCards(count, types = []) {
  const cardsPerType = Math.ceil(count / types.length);
  const cards = types.reduce((cards, type) => {
    for(let i = 0; i < cardsPerType; i += 1) {
      cards.push({
        type,
        name: `${type}-${i}`,
      });
    }
    return cards;
  }, []);
  return cards.splice(0, count);
}

export default function MockDeckFactory() {
  const drawCards = createCards(50, ['ally', 'attachment', 'event']);
  const heroes = createCards(3, ['hero']);
  return {
    allCards: drawCards.concat(heroes),
    drawCards: drawCards,
    heroes: heroes,
  };
}
