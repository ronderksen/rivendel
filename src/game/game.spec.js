import Game from './game';
import createMockDetails from './mocks/details';

jest.mock('./game-pipeline');
import GamePipeline from './game-pipeline';

GamePipeline.mockImplementation(() => ({
  pipeline: [],
  queue: [],
  initialize: jest.fn(),
  queueStep: jest.fn(),
  continue: jest.fn(),
}));

jest.mock('./player');
import Player from './player';
import MockDeckFactory from './test-helpers/MockDeckFactory';
const preparedDeck = MockDeckFactory();

Player.mockImplementation((id, user, isOwner, game) => ({
  id,
  user,
  name: user.username,
  isOwner,
  game,
  left: false,
  preparedDeck,
  initialize: jest.fn(),
}));

jest.mock('./spectator');
import Spectator from './spectator';
Spectator.mockImplementation((id, user, game) => ({
  constructor: Spectator,
  id,
  user,
  game,
  left: false,
  initialize: jest.fn(),
}));

jest.mock('./scenario');
import Scenario from './scenario';
Scenario.mockImplementation(() => ({
  getEncounterSets: jest.fn().mockReturnValue([]),
  getQuestCards: jest.fn().mockReturnValue([]),
}));
jest.mock('./phases/setup-phase');
// jest.mock('./phases/simple-step');

describe('Class: Game', () => {
  const RealDate = global.Date;

  beforeEach(() => {
    const constantDate = new Date('2017-11-28T20:26:42.098Z');
    global.Date = class extends Date {
      constructor() {
        super();
        return constantDate;
      }
    };
  });

  afterEach(() => {
    global.Date = RealDate;
  });

  it('should instantiate given correct details', () => {
    const game = new Game(createMockDetails());
    expect(game).toBeInstanceOf(Game);
    expect(game).toMatchSnapshot();

    expect(Player).toHaveBeenCalledTimes(2);
    expect(Spectator).toHaveBeenCalledTimes(1);
    expect(game.scenario.getEncounterSets).toHaveBeenCalledTimes(1);
    expect(game.scenario.getQuestCards).toHaveBeenCalledTimes(1);
  });

  it('should get an array with all players', () => {
    const game = new Game(createMockDetails());
    const players = game.getPlayers();
    expect(players).toHaveLength(2);
    const names = players.map(player => player.user.username);
    expect(names).toEqual(['chevalric', 'darius']);
  });

  it('should initialize a game', () => {
    const game = new Game(createMockDetails());
    game.initialize();

    game.getPlayers().forEach(player => {
      expect(player.initialize).toHaveBeenCalledTimes(1);
      expect(player.preparedDeck).toEqual(preparedDeck);
    });

    expect(game.pipeline.initialize).toHaveBeenCalledTimes(1);
    expect(game.playStarted).toBe(true);
    expect(game.round).toBe(0);
    expect(game.pipeline.continue).toHaveBeenCalledTimes(1);
  })
});
