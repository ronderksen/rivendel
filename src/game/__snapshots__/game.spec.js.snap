// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class: Game should instantiate given correct details 1`] = `
GameEngine {
  "_events": Object {},
  "_eventsCount": 0,
  "_maxListeners": 0,
  "abilityCardStack": Array [
    Object {
      "card": null,
      "source": "framework",
      "stage": "framework",
    },
  ],
  "abilityWindowStack": Array [],
  "allowSpectators": true,
  "cancelPromptUsed": false,
  "createdAt": 2017-11-28T20:26:42.098Z,
  "difficulty": "easy",
  "domain": null,
  "encounterCards": Array [],
  "id": "test-id001",
  "name": "Test game",
  "owner": "chevalric",
  "password": "",
  "pipeline": GamePipeline {
    "pipeline": Array [],
    "queue": Array [],
  },
  "playStarted": false,
  "playerCards": Object {},
  "playerHeroes": Object {},
  "playersAndSpectators": Object {
    "chevalric": Player {},
  },
  "questCards": Array [],
  "savedGameId": null,
  "scenario": Scenario {},
  "scenarioId": "01001",
  "skipPhase": Object {},
  "started": false,
}
`;
